README.md
Nahuel Scotti - Portfolio
===

Here is what I have been working on.

[![Build Status](https://gitlab.com/singuerinc-works/com.singuerinc.www/badges/master/build.svg)](https://gitlab.com/singuerinc-works/com.singuerinc.www/commits/master)
[![Coverage Report](https://gitlab.com/singuerinc-works/com.singuerinc.www/badges/master/coverage.svg)](https://gitlab.com/singuerinc-works/com.singuerinc.www/commits/master)

This is my portfolio, a place that displays some work and experiments I have
been working in the last few years.

![](screenshot.png)

## Installation

Althougth it is possible to install all dependencies on a local environment
(Node, Ruby, yarn, etc), the setup is highly optimized to work with Docker.

1. Install [Docker](https://www.docker.com/products/docker)
2. Clone the project
	
	```sh
	git clone git@gitlab.com:singuerinc-works/com.singuerinc.www.git
	cd www
	```
3. Start the Docker container

	```sh
	docker-compose up jekyll
	```

4. Open a browser and navigate to `http://localhost:4000`

## Usage example

All versions are available as Docker images.
If your idea is serve the website, start it by executing these commands:

```sh
docker login registry.gitlab.com
docker run -p 8080:80 registry.gitlab.com/singuerinc-works/com.singuerinc.www
open 'http://localhost:8080'
```

## Development setup

The project is generated with [Jekyll](https://jekyllrb.com/)

The Jekyll Docker container is in charge of compile the project every
time a change is made.

A compiled version could be found in the `public` folder.

The Jekyll container also relay on another container called `webpack` that
compiles the Javascript code every time a change is done to a js file.

### Tests

#### Unit test - AVA/Javascript
Test all the Javascript code.

```sh
docker-compose run webpack yarn run test
```

#### Links tests
Checks if all the links (internal and extenal) are working.

```sh
docker-compose run tests-links
```

## Make a change

Usually, to make a change follow the next workflow:

1. Create an Issue on GitLab
2. Create a new branch for that Issue
3. Check out the branch
4. Make your changes
5. Commit and push the changes
6. The CI build, tests, and release the feature branch to [http://test.www.singuerinc.com.surge.sh/](http://test.www.singuerinc.com.surge.sh/)
7. If the status is green, merge into `master` from GitLab.
8. The CI build, tests, and release the master branch to [http://staging.www.singuerinc.com.surge.sh/](http://staging.www.singuerinc.com.surge.sh/)
9. If the status is green, tag the `master` in GitLab.
10. The CI build, tests, and release the tag to [https://www.singuerinc.com/](https://www.singuerinc.com/)

## Release History

See [https://gitlab.com/singuerinc-works/com.singuerinc.www/blob/master/CHANGELOG](https://gitlab.com/singuerinc-works/com.singuerinc.www/blob/master/CHANGELOG)

## Meta

Nahuel Scotti – [@singuerinc](https://twitter.com/singuerinc) – nahuel.scotti@gmail.com

[https://github.com/singuerinc/](https://github.com/singuerinc/)


CHANGELOG
21.3.0
==================

  * update animejs to version 2
  * change to node:alpine-8 images w/yarn included
  * optimize some lighthouse stuff
 
21.2.0
==================

  * optimize images
  * limit to 2 the number of images to pre-load
 
21.1.0
==================

  * build js for production without sourcemaps
 
21.0.0
==================

  * Refactor js: Use TypeScript
 
20.1.3
==================

  * new acme keys
 
20.1.2
==================

  * fix: add deps
 
20.1.1
==================

  * update acme content
 
20.1.0
==================

  * dynamic acme file generation
 
20.0.1
==================

  * Remove 'function test' from CI
 
20.0.0
==================

  * Add Subway project
 
19.0.0
==================

  * Add "Tomeito-app" project
 
18.2.1
==================

  * acme file
 
18.3.0
==================

  * acme file


18.2.0 / 2017-03-03
==================

  * Add lint to staging and production environment
  * Merge branch '137-update-readme-file' into 'master'
  * Merge branch '134-use-airbnb-style-js' into 'master'
  * Add authors file
  * Update CHANGELOG markup
  * Merge branch '138-use-absolute-path-in-favicon' into 'master'
  * Add airbnb eslint config, fix js files

18.1.0 / 2017-03-01
==================

  * Merge branch '138-use-absolute-path-in-favicon' into 'master'
  * Use absolute path on favicon

18.0.0 / 2017-02-26
==================

  * Add unit tests to javascript app

17.1.0 / 2017-01-22
==================

  * Use Open Color palette

17.0.0 / 2016-12-22
==================

  * Add Columna Colors project

16.7.0 / 2016-12-21
==================

  * Add metas to projects

16.6.0 / 2016-12-21
==================

  * Add Atrapalo Revivelo screenshots

16.5.0 / 2016-12-21
==================

  * Fix 404 page javascript error

16.4.1 / 2016-12-21
==================

  * Fix SITE_URL protocol

16.4.0 / 2016-12-21
==================

  * Fix 404 styles

16.3.0 / 2016-12-20
==================

  * Fix #128 / Loading bar display with blue color until loads css

16.2.0 / 2016-12-20
==================

  * Fix back button and Turbolinks cache problem

16.1.0 / 2016-12-19
==================

  * Add new ACME file

16.0.0 / 2016-12-19
==================

  * Add test for category field

15.0.0 / 2016-11-16
==================

  * Add "Bi" project

14.0.0 / 2016-11-16
==================

  * Sitemap: Rename to map
  * Add Project title to page title
  * Change CI builds to GIT_STRATEGY: none when needed

13.0.0 / 2016-11-16
==================

  * Renew SSL Certificate
  * Change title color (gradient background)
  * Remove Evax "Adapt" project
  * Better project's descriptions
  * Add social icons in home page
  * Remove project's client from home page
  * Load new page with Turbolinks
  * Add next project button
  * Fix singuerinc link to home page

12.0.0 / 2016-09-29
==================

  * Add enter and leave transitions to pages
  * Add next project button
  * Add some animations to the project's page
  * Fix singuerinc link to home page

11.0.0 / 2016-09-22
==================

  * Fix Images are not big enough
  * Add site map page
  * Remove <br/>'s
  * Implement progress bar on loading
  * Add role to project profiles
  * Award box is not vertically centered
  * git commit sha in html doc
  * Meta tags according to the page

10.0.0 / 2016-09-15
==================

  * More semantic tags for project title, etc
  * Minimize CSS
  * Add Social icons
  * Add humans.txt

9.1.0 / 2016-09-11
==================

  * Create Twitter Card

9.0.0 / 2016-09-11
==================

  * Fix Black placeholders during the images loading should be white or transparent
  * Fix Related buttons are too small for mobile
  * Fix Images that are 50% tall are displayed incorrectly in mobile
  * Fix Style inside a div is not permitted
  * Test in devices

8.0.0 / 2016-09-10
==================

  * optimize images for desktop
  * rewrite app.js

7.1.0 / 2016-09-09
==================

  * Add Home button in the sidebar

7.0.0 / 2016-09-08
==================

  * Add Evax "Estudio Risa" project
  * Update Readme file with test command
  * Remove "scoped" from <style> tag
  * Fix Docker test command
  * Fix Project title should not be duplicated
  * Remove header variable "host"

6.0.0 / 2016-08-28
==================

  * Fix Google Analytics is sending statistics during the tests
  * Use anchors in gitlab-ci.yml
  * Exclude robots from test and staging environment
  * Add CNAME for each environment
  * Test files loaded
  * Change $TRAVIS_BUILD_DIR variable to $CI_PROJECT_DIR in .gitlab-ci.yml

5.2.2 / 2016-08-26
==================

  * Fix Docker tag

5.2.1 / 2016-08-26
==================

  * Fix fix only latest docker image is created !40

5.2.0 / 2016-08-25
==================

  * Fix "Google Bot doesn't see anything"
  * Optimize assets after PageSpeed Insights report
  * Exclude index.js from compiled public folder
  * Test before each commit (via git hook)
  * Fix 404 white text and styles
  * Test 404 case

5.1.0 / 2016-08-24
==================

  * Add 200 status test
  * Add Google Analytics test on all pages
  * Update Google Analytics script

5.0.0 / 2016-08-24
==================

  * Add Functional tests
  * Change About title
  * htmlproofer now checks external links
  * lock jekyll image to 3.2.1 tag

4.0.0 / 2016-08-21
==================

  * Add Facebook App ID
  * Add Facebook Admin ID
  * Fix absolute paths (bundle.js, analytics.js)
  * Remove unused images
  * Max width for project images
  * Audi Driving Experience: Hi-res image
  * FCB Som Un: Hi-res image
  * Audi A1 Me Gusta: Hi-res image

3.3.0 / 2016-08-19
==================

  * Change og:image to Ivy Park

3.2.2 / 2016-08-18
==================

  * Change Audi "Driving Experience" url
  * Change FCB "Som Un" url
  * Change MDM "Names not numbers" url

3.2.1 / 2016-08-15
==================

  * Add Skoda website url

3.2.0 / 2016-08-14
==================

  * Change Metadata id
  * Awards badge

Portfolio.js
/* global window, setTimeout, Image, document, Turbolinks */
/* eslint-disable class-methods-use-this */
import NProgress from 'nprogress';
import { anime } from './anime';

/**
 * The Portfolio is the entry point for every request the website does.
 * When the page is loaded, the application resolve which page should be
 * loaded and ask the portfolio to load the correct content.
 * @class Portfolio
 */
class Portfolio {

  constructor() {
    const STYLE = this.getComputedStyle(document.querySelector('html'));

    this.isRetina = window.devicePixelRatio > 1;
    this.isMobile = (parseInt(STYLE.getPropertyValue('width'), 10)) < 768;
  }

  /**
   * Adds a listener to each element.
   * When the user clicks in one of those elements,
   * then the pages fadeout and navigate to the next page.
   * @private
   * @param {string} elements - The selector string.
   * @returns {void}
   */
  whenClickExit(elements) {
    document.querySelectorAll(elements).forEach((element) => {
      element.addEventListener('click', (event) => {
        event.preventDefault();

        anime({
          targets: '.content',
          opacity: [1, 0],
          duration: 700,
          easing: 'easeInOutExpo',
          complete: () => {
            Turbolinks.visit(event.target.getAttribute('href'), { action: 'advance' });
          },
        });
      });
    });
  }

  /**
   * _onLoad acts every time an image is loaded.
   * It display the image and check if all the images
   * are loaded, in that case calls _onIndexReady that
   * execute an animation.
   * @private
   * @param {string} postId - The post identifier.
   * @param {number} totalImages - The amount of images to load.
   * @param {string} src - The path to the image.
   * @returns {void}
   */
  onLoad(postId, totalImages, src) {
    const tag = document.querySelector(`.post-image.${postId}`);

    tag.style.backgroundImage = `url(${src})`;

    this.totalImagesLoaded += 1;
    NProgress.set(this.totalImagesLoaded / totalImages);
    if (this.totalImagesLoaded === totalImages) {
      const waitBeforeReady = 200;

      setTimeout(() => {
        NProgress.done(true);
        this.onIndexReady();
      }, waitBeforeReady);
    }
  }

  /**
   * Starts the load of all images in the portfolio.
   * @param {array} posts - The list of posts to show in the index.
   * @returns {void}
   */
  loadIndex(posts) {
    this.whenClickExit('h1 a, .sidebar nav > ul > li > a, .sidebar-mobile ul li a, ul.posts li > .w-link');
    this.totalImagesLoaded = 0;
    posts.forEach((post, index) => {
      let filename;

      if (this.isMobile && !this.isRetina) {
        filename = `${post.image}-md.jpg`;
      } else {
        filename = `${post.image}.jpg`;
      }

      const src = `./img/home/${filename}`;

      if (index < 4) {
        const image = new Image();

        image.onload = () => this.onLoad(post.id, posts.length, image.src);
        image.src = src;
      } else {
        this.onLoad(post.id, posts.length, src);
      }
    });
  }

  /**
   * Animates the previous and the next project in the page.
   * @private
   * @returns {void}
   */
  showPrevAndNextProjects() {
    anime({
      targets: '.page .prev-next-project',
      begin: animation => animation.animatables[0].target.classList.remove('hide'),
      opacity: [0, 1],
      duration: 2000,
      delay: 150,
      easing: 'easeInOutExpo',
    });
  }

  /**
   * Animates the title of the page.
   * @private
   * @returns {void}
   */
  showTitle() {
    anime({
      targets: '.page .title',
      begin: animation => animation.animatables[0].target.classList.remove('hide'),
      opacity: [0, 1],
      duration: 1500,
      easing: 'easeInOutExpo',
    });
  }

  /**
   * When all images in the home page are loaded an animation is played.
   * @private
   * @returns {void}
   */
  onIndexReady() {
    anime({
      targets: '.pre.hide',
      begin: animation => animation.animatables[0].target.classList.remove('hide'),
      opacity: [0, 1],
      duration: 1500,
    });

    document.querySelectorAll('ul.posts li').forEach((element) => {
      element.classList.remove('hide');
      const link = element.querySelector('.w-link');

      link.style.backgroundColor = 'black';
    });

    anime({
      targets: '.posts li:nth-child(-n+4)',
      translateY: [400, 0],
      opacity: [0, 1],
      duration: 1500,
      easing: 'easeInOutExpo',
      delay: (element, index) => 250 * index,
    });
  }

  /**
   * Gets the computed style of an element.
   * @private
   * @param {HTMLElement} elem - The element you want to compute.
   * @returns {Object} The computed style.
   */
  getComputedStyle(elem) {
    if (elem.ownerDocument.defaultView.opener) {
      return elem.ownerDocument.defaultView.getComputedStyle(elem, null);
    }
    return window.getComputedStyle(elem, null);
  }

  /**
   * Executed when the about page is displayed.
   * @returns {void}
   */
  loadAbout() {
    this.whenClickExit('h1 a, .sidebar nav > ul > li > a, .sidebar-mobile ul li a');

    NProgress.inc();
    const image = new Image();

    image.onload = () => {
      NProgress.done(true);
      this.showTitle();

      anime({
        targets: '.content > p.text, .content p img',
        translateY: [100, 0],
        translateX: () => [anime.random(0, 500), 0],
        opacity: [0, 1],
        duration: 1500,
        easing: 'easeInOutExpo',
        delay: (element, index) => 50 * index * Math.random(),
      });

      anime({
        targets: '.content blockquote',
        begin: animation => animation.animatables[0].target.classList.remove('hide'),
        opacity: [0, 1],
        duration: 1500,
        delay: 6000,
        easing: 'easeInOutExpo',
      });

      setTimeout(() => document.querySelectorAll('.content p img, .content > p.text').forEach(e => e.classList.remove('hide')), 0);
    };

    image.src = `${document.querySelector('.page .content img').getAttribute('src')}`;
  }

  /**
   * Executed when the site-map page is displayed.
   * @returns {void}
   */
  loadSiteMap() {
    this.whenClickExit('h1 a, .sidebar nav > ul > li > a, .sidebar-mobile ul li a, .site-map a');
    this.showTitle();

    setTimeout(() => document.querySelectorAll('.site-map li').forEach(e => e.classList.remove('hide')), 1);

    anime({
      targets: '.site-map li',
      translateY: [100, 0],
      translateX: () => [anime.random(0, 500), 0],
      opacity: [0, 1],
      duration: 1500,
      easing: 'easeInOutExpo',
      delay: (element, idx) => idx * 25,
    });

    anime({
      targets: '.content blockquote',
      begin: animation => animation.animatables[0].target.classList.remove('hide'),
      opacity: [0, 1],
      duration: 1500,
      delay: 6000,
      easing: 'easeInOutExpo',
    });
  }

  /**
   * Executed when the 404 page is displayed.
   * @returns {void}
   */
  load404() {
    this.whenClickExit('h1 a, .sidebar nav > ul > li > a, .sidebar-mobile ul li a, .site-map a');
    this.showTitle();

    anime({
      targets: '.content p',
      translateY: [100, 0],
      translateX: [anime.random(0, 500), 0],
      opacity: [0, 1],
      duration: 1500,
      easing: 'easeInOutExpo',
    });
  }

  /**
   * Executed when the 404 page is displayed.
   * @returns {void}
   */
  loadProject() {
    this.whenClickExit('h1 a, .sidebar nav > ul > li > a, .sidebar-mobile ul li a, .prev-next-project li a, .project-page .related-post a');

    NProgress.inc();
    const image = new Image();

    image.onload = () => {
      NProgress.done(true);
      this.showTitle();

      const targets = [
        '.project-page .project-title',
        '.project-page .image',
        '.project-page .info tr',
        '.project-page .project-content p',
        '.project-page .project-content .video-wrapper',
        '.project-page .project-content ul',
        '.project-page .share-title',
        '.project-page .share-post',
        '.project-page hr',
        '.project-page .related-title',
        '.project-page .related-post',
      ];

      setTimeout(() => {
        const content = document.querySelector('.content');

        content.classList.remove('hide');
      }, 1);

      anime({
        targets: '.content',
        opacity: [0, 1],
        easing: 'easeInOutExpo',
      });

      anime({
        targets: '.project-page .prev-next-project',
        opacity: [0, 1],
        duration: () => 1500,
        easing: 'easeInOutExpo',
      });

      anime({
        targets,
        translateY: [100, 0],
        translateX: () => [anime.random(100, 500), 0],
        opacity: [0, 1],
        duration: () => 1500,
        easing: 'easeInOutExpo',
      });
    };

    image.src = `${document.querySelector('.project-page .image').getAttribute('src')}`;
  }
}

export default Portfolio;
